engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java

```diff
@@ -399,18 +423,63 @@ public Object lookupJavaClass(String className) {
       List<String> nestedClassPart =
           i < items.size() - 1 ? items.subList(i + 1, items.size()) : List.of();
       try {
-        Object hostSymbol = environment.lookupHostSymbol(pkgName + "." + curClassName);
+        var hostSymbol = lookupHostSymbol(pkgName, curClassName);
         if (nestedClassPart.isEmpty()) {
           return hostSymbol;
         } else {
-          return getNestedClass(hostSymbol, nestedClassPart);
+          var fullInnerClassName = curClassName + "$" + String.join("$", nestedClassPart);
+          return lookupHostSymbol(pkgName, fullInnerClassName);
         }
-      } catch (RuntimeException ignored) {
+      } catch (RuntimeException | InteropException ex) {
+        logger.log(Level.WARNING, null, ex);
       }
     }
     return null;
   }
 
+  private Object lookupHostSymbol(String pkgName, String curClassName)
+  throws UnknownIdentifierException, UnsupportedMessageException {
+    if (findGuestJava() == null) {
+      return environment.lookupHostSymbol(pkgName + "." + curClassName);
+    } else {
+      return InteropLibrary.getUncached().readMember(findGuestJava(), pkgName + "." + curClassName);
+    }
+  }
+
+  private Object guestJava = this;
+
+  @TruffleBoundary
+  private Object findGuestJava() throws IllegalStateException {
+    if (guestJava != this) {
+      return guestJava;
+    }
+    guestJava = null;
+    var envJava = System.getenv("ENSO_JAVA");
+    if (envJava == null) {
+      return guestJava;
+    }
+    if ("espresso".equals(envJava)) {
+      var src = Source.newBuilder("java", "<Bindings>", "getbindings.java").build();
```

Comment:
where is this coming from?

Reply:
This is a standard way to deal with [Espresso](https://github.com/oracle/graal/blob/764fc7e1dd72521ea70eaeed165205f902c6dc3b/espresso/src/com.oracle.truffle.espresso/src/com/oracle/truffle/espresso/nodes/commands/GetBindingsNode.java#L31).